Process data uri images synchronously

jonmmease opened the pull request:
This PR is the result of some time spent debugging an odd behavior I was seeing when using plotly.js layout images to display interactive results generated by [Datashader](http://datashader.org/).

The basic idea is that every time the viewport of a plot changes (pan or zoom), a new image is generated by datashader to be displayed for that viewport.  By encoding these images a data uris, I've been able to display them pretty successfully using `layout.image` objects.

But one problem has been that sometimes the images appear to jump right before they are updated.  Here's what that looks like.

![datashader_jump](https://user-images.githubusercontent.com/15064365/62414978-bf6ed100-b5f0-11e9-8cd9-a95f281b69a3.gif)

After playing with this for a while, I realized that what's happening is that the image position (with the old image) is getting updated a moment before the new image is displayed.  Here's a pure JavaScript example that shows the problem a little more clearly. (gist: https://gist.github.com/jonmmease/d2793251fdb1f370e483d54ca6561690).  In this snippet, each time the viewport changes, `Plotly.react` is used to update the image with a new `source` and a new position, where the position is always the bottom right corner of the viewport.

![datauri_before](https://user-images.githubusercontent.com/15064365/62414879-6e121200-b5ef-11e9-8fb0-44908d3203c3.gif)

Notice how the old image is moved to the bottom right corner before the new image is displayed.

After digging into plotly.js a bit, I realized that the reason this is happening is that the displayed image is updated in an `img.onload` callback that is executed after the image has loaded.   This callback then builds and saves the data uri form of the image using a canvas element.

This PR makes a small adjustment to bypass the `onload` callback for images that are specified as data uri's already.  In this case we can use the `layout.image.source` string directly.  With this change, images specified as data uris will update at the same time that their position is updated. This change is in commit 536c90e. Commit 0a397c4 contains the indentation that I omitted from 536c90e in order to make a cleaner diff of the logic changes.

Here is what the example above looks like in this PR:
![datauri_after](https://user-images.githubusercontent.com/15064365/62414881-723e2f80-b5ef-11e9-916a-8cd5376cc00c.gif)

Let me know if there are any specific tests you'd like to see for this change.  I did notice that there is already a data uri image being tested by the `layout_image` mock.

jonmmease commented:
I'm not sure what's going on with the tests.
```
    ✗ @gl should update the scene camera - orthographic case (3494ms)
	Error: Timeout - Async callback was not invoked within 3000ms (set by jasmine.DEFAULT_TIMEOUT_INTERVAL)
```

And another timeout.  I already tried restarting them once with no change. Any ideas?

etpinard commented:
> And another timeout. I already tried restarting them once with no change. Any ideas?

Tests are passing now.

etpinard commented on src/components/images/draw.js:
```diff
@@ -84,31 +84,37 @@ module.exports = function draw(gd) {
-                var canvas = document.createElement('canvas');
-                canvas.width = this.width;
-                canvas.height = this.height;
+            if(d.source && d.source.slice(0, 5) === 'data:') {
```
`d.source` should always be a string at this stage, so you could use:

```js
if(d.source.indexOf('data:') === 0) {
  // ...
}
```

instead.

jonmmease replied:
👍

alexcjohnson replied:
`d.source` can be large though, isn't `slice(0, 5)` going to be a lot faster than `indexOf`?

jonmmease replied:
That's a good point @alexcjohnson. @etpinard, anything you don't like about `slice` here?

etpinard replied:
It's not obvious to me why `slice` would be faster than `indexOf() === 0`. Has anyone tested that out on jsperf before?

jonmmease replied:
My thinking was that If the string starts with `'data:'`, then `indexOf()` would terminate right away. But if not, `indexOf()` would search the full string and not bail out until it reaches the end, whereas slice never needs to look past the first 5 characters.  If the string doesn't start with `data:`, then I guess it should be a URL which hopefully isn't terribly long. So it might not matter either way.

alexcjohnson replied:
I guess in as far as the two options we care about are a short string starting `http`
and a long string starting `data:`, `indexOf` should be fine. The case to worry about would be a long string that doesn't start `data:`... but we don't expect that and wouldn't care if it's slow.

jsperf shows a win for `indexOf` https://jsperf.com/slice-vs-indexof-acj but I don't trust it, there's no way it could actually do `indexOf` in 2 clock cycles, compiler must have turned it into a constant. This isn't happening in a loop so either option is fast enough to be not worth worrying about.

etpinard commented on src/components/images/draw.js:
```diff
@@ -84,31 +84,37 @@ module.exports = function draw(gd) {
-                canvas.width = this.width;
-                canvas.height = this.height;
+            if(d.source && d.source.slice(0, 5) === 'data:') {
+                img.src = d.source;
```
I see that you're setting `img.src` here even though the `Image` instance isn't used in this case. That's not a bad idea as this makes the early return a few lines above

```js
        if(this.img && this.img.src === d.source) {
            return;
        }
```

work.

But setting `img.src` spins off the `img.onload`, `img.onerror`, etc event chain. I'm not sure how if that can lead to performance losses, but it would be easy to stash the `d.source` data in something like `img._src` instead to skip potential `img.onload` / `img.onerror` work.

etpinard replied:
We could even go one step further by

- if `source` is a data uri, do not wrap logic in a `Promise` (and do not append `gd._promises`)
- if `source` is a data uri, do not instantiate `Image`
- use `thisImage.attr('xlink:href')` as the "stash" value instead of `img.src`

jonmmease replied:
Yeah, that's a good point. `img._src` sounds good.

etpinard commented:
Thanks @jonmmease - this works!

Before merging this, it would be cool to add one test checking that layout images with data uri as `source` do not call `canvas.getContext` and/or `canvas.toDataURL`. Here's one way we currently use to spy on canvas methods:

https://github.com/plotly/plotly.js/blob/ffa6e3468b592e45db2a00bb65805a92fd59c4fb/test/jasmine/tests/heatmap_test.js#L735-L741

etpinard commented:
Oh and @jonmmease , I put this under the `1.50.0` milestone. This _could_ be released as part of a patch if urgent, but I would prefer releasing this in a minor. I hope you don't mind.

jonmmease commented:
Thanks @etpinard,  yeah that test approach make sense. I'll give it a go!

Not urgent, 1.50.0 is fine. Thanks for the quick review!

jonmmease commented:
Ok, in c310c92 I pulled the data uri image handling outside of the promise. So in this case we don't create the promise, construct an `Image`, or run `onload`.  And I added a `this._imgSrc` property to save off the input image string for later comparison.

In dd585b2, I added a test to make sure that we don't construct any `Image` elements for the data uri case, but we do construct an image and call `toDataURL` for the image url case.

**Note:** I tried to avoid the `setTimeout` call in the test, but I couldn't figure out how to get Jasmine to check that a variable remains undefined, without timing out the test. `expect(newCanvasElement).not.eventually.toBeDefined()`, `expect(newCanvasElement).eventually.not.toBeDefined()`, and `expect(newCanvasElement).eventually.toBeUndefined()` all timed out.

etpinard commented on test/jasmine/tests/layout_images_test.js:
```diff
@@ -290,6 +294,31 @@ describe('Layout images', function() {
+                return element;
+            });
+
+            Plotly.relayout(gd, 'images[0].source', dataUriImage).then(function() {
```
Non-blocking but something like:

```js
        it('should only create canvas if url image', function(done) {
            var originalCreateElement = document.createElement;
            var newCanvasElement;
            spyOn(document, 'createElement').and.callFake(function(elementType) {
                var element = originalCreateElement.call(document, elementType);
                if(elementType === 'canvas') {
                    newCanvasElement = element;
                    spyOn(newCanvasElement, 'toDataURL');
                }
                return element;
            });

            Plotly.relayout(gd, 'images[0].source', dataUriImage)
            .then(function() {
                expect(newCanvasElement).toBeUndefined();
            })
            .then(function() { return Plotly.relayout(gd, 'images[0].source', pythonLogo); })
            .then(function() {
                expect(newCanvasElement).toBeDefined();
                expect(newCanvasElement.toDataURL).toHaveBeenCalledTimes(1);
            })
            .catch(failTest)
            .then(done)
        });
```

should work.

jonmmease replied:
The trouble there is that the first `expect(newCanvasElement).toBeUndefined();` will also pass for a url image, because the `onload` callback won't necessarily have completed yet. At least that's what I concluded was happening when I tried this.

etpinard replied:
> The trouble there is that the first `expect(newCanvasElement).toBeUndefined();` will also pass for a url image, because the `onload` callback won't necessarily have completed yet.

Hmm, that shouldn't happen, the `Plotly.relayout` promise shouldn't resolve before the `onload` callback is executed.

jonmmease replied:
Ok, I'll double check...

jonmmease replied:
Yeah, you're right. Not sure what I was trying before.  I updated the test to use this structure.

etpinard commented:
Very strong PR @jonmmease ! Thanks very much for your help!

We haven't started merging things for `1.50.0`, so I won't give you a `: dancer :` just yet. We'll most likely release 1 or 2 more patch releases in the `1.49.x` series before.

jonmmease commented:
Thanks for the reviews @etpinard, nice to dip my toe into the code base again!
